### PR TITLE
Match dependencies to their respective descriptors after Python scan

### PR DESCRIPTION
**Pull request by Or-Geva**
- [ ] All [tests](https://github.com/jfrog/jfrog-vscode-extension#building-and-testing-the-sources) passed. If this feature is not already covered by the tests, I added new tests.
- [x] I used `npm run format` for formatting the code before submitting the pull request.
-----
Despite the fact that each descriptor contains different dependencies, the Python dependency scan shows the same vulnerable dependencies across all descriptors (setup.py/requirements.txt).

As a result of this PR, each direct dependency and its child are mapped to the right descriptor. In addition, some dependencies, such as pip and Python binaries in the virtual environment, do not belong to any descriptor.
This will result in a new node called `Virtual Environment` that will appear beside all the project descriptors, containing all the tools that belong to the Python virtual environment 

image

### CODE REVIEW

File: src/main/utils/pypiUtils.ts [hunk 1]
```diff
@@ -163,24 +143,141 @@ export class PypiUtils {
...
+        let projectDependencies: PipDepTree[] | undefined;
+        try {
+            projectDependencies = JSON.parse(ScanUtils.executeCmd(pythonPath + ' ' + PypiUtils.PIP_DEP_TREE_SCRIPT + ' --json-tree').toString());
+        } catch (error) {
+            logManager.logError(<any>error, true);
+        }
+        return projectDependencies;
+    }
+
+    public static async descriptorsToDependencyTrees(
```
Comment (yahavi) [inline src/main/utils/pypiUtils.ts:166]: Let's add some JSDoc

File: src/main/utils/pypiUtils.ts [hunk 2]
```diff
@@ -163,24 +143,141 @@ export class PypiUtils {
...
+            checkCanceled();
+            treesManager.logManager.logMessage(`Analyzing '${descriptor.fsPath}' file`, 'INFO');
+            let root: PypiTreeNode = new PypiTreeNode(descriptor.fsPath, parent);
+            root.refreshDependencies(this.filterDescriptorDependencies(descriptor.fsPath, pipDepTree, projectName));
+            trees.push(root);
+        }
+        return [trees, pipDepTree];
+    }
+
+    public static async workspaceToDependencyTree(
```
Comment (yahavi) [inline src/main/utils/pypiUtils.ts:185]: Let's add some JSDoc

File: src/main/utils/pypiUtils.ts [hunk 3]
```diff
@@ -163,24 +143,141 @@ export class PypiUtils {
...
+        parent: DependenciesTreeNode,
+        projectsToScan: ProjectDetails[]
+    ) {
+        let root: VirtualEnvPypiTree = new VirtualEnvPypiTree(virtualEnvPath, workspace.uri.fsPath, parent);
+        root.refreshDependencies(pipDepTree);
+        parent.children.push(root);
+        projectsToScan.push(root.projectDetails);
+    }
+
+    private static filterDescriptorDependencies(descriptorPath: string, pipDepTree: PipDepTree[], projectName?: string): PipDepTree[] {
```
Comment (yahavi) [inline src/main/utils/pypiUtils.ts:198]: Let's add some JSDoc

File: src/main/utils/pypiUtils.ts [hunk 4]
```diff
@@ -163,24 +143,141 @@ export class PypiUtils {
...
+        const dependencies: Map<string, string | undefined> = isSetupPy
+            ? this.getSetupPyDirectDependencies(descriptorPath)
+            : this.getRequirementsTxtDirectDependencies(descriptorPath);
+        if (!dependencies) {
+            return pipDepTree;
+        }
+        return this.filterDependencies(dependencies, pipDepTree, false, projectName);
+    }
+
+    public static filterDependencies(
```
Comment (yahavi) [inline src/main/utils/pypiUtils.ts:209]: Let's add some JSDoc

File: src/main/utils/pypiUtils.ts [hunk 5]
```diff
@@ -163,24 +143,141 @@ export class PypiUtils {
...
+    }
+
+    private static isVersionsEqual(dependencyFromPipDepTree: PipDepTree, depFromDescriptor: string, isSetupPy: boolean): boolean {
+        if (isSetupPy) {
+            return dependencyFromPipDepTree.required_version === depFromDescriptor;
+        }
+        return depFromDescriptor.endsWith(dependencyFromPipDepTree.installed_version);
+    }
+
+    private static scanForProjectName(descriptors: vscode.Uri[]): string | undefined {
```
Comment (yahavi) [inline src/main/utils/pypiUtils.ts:242]:
```suggestion
    private static getProjectName(descriptors: vscode.Uri[]): string | undefined {
```

File: src/main/utils/pypiUtils.ts [hunk 6]
```diff
@@ -163,24 +143,141 @@ export class PypiUtils {
...
+        let installReqRegex: RegExp = /install_requires=\[(.*)\]/gms;
+        const match: RegExpExecArray | null = installReqRegex.exec(content);
+        if (!match) {
+            return new Map<string, string | undefined>();
+        }
+        return this.matchPythonDependencies(match[1]);
+    }
+
+    private static matchPythonDependencies(rawDependencies: string) {
+        let packageRegex: RegExp = /([\w\d\-.]+)\s*(?:\[.*\])?\s*((?:\s*(?:[<>]=?|!=|===?|~=)\s*[\d\w*-.]+,?)*)/gms;
```
Comment (yahavi) [inline src/main/utils/pypiUtils.ts:267]: Let's export all Regex patterns in this file to class constants

Comment (yahavi) [inline src/main/utils/pypiUtils.ts:267]: Isn't the `.` already contain `\w` and `\d`?
```suggestion
        let packageRegex: RegExp = /([-.]+)\s*(?:\[.*\])?\s*((?:\s*(?:[<>]=?|!=|===?|~=)\s*[*-.]+,?)*)/gms;
```

**Review by yahavi — CHANGES_REQUESTED** [review]

File: src/main/utils/pypiUtils.ts [hunk 7]
```diff
@@ -163,24 +143,141 @@ export class PypiUtils {
...
+        while ((dependencyMatch = packageRegex.exec(rawDependencies)) !== null) {
+            const [, name, version] = dependencyMatch;
+            dependencies.set(name.toLowerCase(), version);
+        }
+        return dependencies;
+    }
+
+    private static searchProjectName(setupPyFile: string) {
+        const content: string = fs.readFileSync(setupPyFile, 'utf8');
+        const nameRegex: RegExp = new RegExp(`name=\\s*(?:"|')(.*)(?:"|')`, 'g');
```
Comment (yahavi) [inline src/main/utils/pypiUtils.ts:279]: Please reconsider this Regex. It doesn't work in all cases.

File: src/main/treeDataProviders/issuesTree/issuesTreeDataProvider.ts [hunk 8]
```diff
@@ -290,30 +299,40 @@ export class IssuesTreeDataProvider implements vscode.TreeDataProvider<IssuesRoo
 
         progressManager.startStep('🔎 Scanning for issues', graphSupported ? 2 * descriptorsCount + 1 : 1);
         let scansPromises: Promise<any>[] = [];
-        scansPromises.push(AnalyzerUtils.runEos(workspaceData, root, workspaceDescriptors, this._scanManager, progressManager));
+        scansPromises.push(AnalyzerUtils.runEos(scanResults, root, workspaceDescriptors, this._scanManager, progressManager));
         // Dependency graph scan and applicability scan for each descriptor
         if (graphSupported) {
```
[the remark this answers is not in the document]
Comment (Or-Geva) [reply src/main/treeDataProviders/issuesTree/issuesTreeDataProvider.ts:304]: I will open a new PR for this change.

File: src/main/treeDataProviders/issuesTree/issuesTreeDataProvider.ts [hunk 9]
```diff
@@ -324,15 +343,15 @@ export class IssuesTreeDataProvider implements vscode.TreeDataProvider<IssuesRoo
         let scansPromises: Promise<any>[] = [];
```
[the remark this answers is not in the document]
Comment (Or-Geva) [reply src/main/treeDataProviders/issuesTree/issuesTreeDataProvider.ts:343]: Removed

**Pull request merged**